File: pfm-js/src/export.ts
/**
 * PFM Export — Turn parser and training data exporters.
 *
 * Converts .pfm conversation files into fine-tuning JSONL formats:
 *   - OpenAI  (messages array with system/user/assistant roles)
 *   - Alpaca  (instruction/input/output per turn pair)
 *   - ShareGPT (conversations array with human/gpt roles)
 */

import { readFileSync, readdirSync, statSync } from 'node:fs';
import { join, extname } from 'node:path';
import { parse } from './parser.js';
import type { PFMDocument } from './types.js';

// ---------------------------------------------------------------------------
// Turn parser
// ---------------------------------------------------------------------------

export interface Turn {
  role: 'user' | 'assistant';
  content: string;
}

export function parseTurns(text: string): Turn[] {
  const turns: Turn[] = [];
  if (!text || !text.trim()) return turns;

  const blocks = text.split('\n\n');
  let currentRole: 'user' | 'assistant' | null = null;
  let currentLines: string[] = [];

  for (const block of blocks) {
    const stripped = block.trim();
    if (!stripped) continue;

    if (stripped.startsWith('User:')) {
      if (currentRole !== null) {
        turns.push({ role: currentRole, content: currentLines.join('\n\n').trim() });
      }
      currentRole = 'user';
      currentLines = [stripped.slice('User:'.length).trim()];
    } else if (stripped.startsWith('Assistant:')) {
      if (currentRole !== null) {
        turns.push({ role: currentRole, content: currentLines.join('\n\n').trim() });
      }
      currentRole = 'assistant';
      currentLines = [stripped.slice('Assistant:'.length).trim()];
    } else if (stripped.startsWith('Agent:')) {
      if (currentRole !== null) {
        turns.push({ role: currentRole, content: currentLines.join('\n\n').trim() });
      }
      currentRole = 'assistant';
      currentLines = [stripped.slice('Agent:'.length).trim()];
    } else {
      if (currentRole !== null) {
        currentLines.push(stripped);
      } else {
        currentRole = 'assistant';
        currentLines = [stripped];
      }
    }
  }

  if (currentRole !== null) {
    turns.push({ role: currentRole, content: currentLines.join('\n\n').trim() });
  }

  return turns;
}

// ---------------------------------------------------------------------------
// Metadata extractor
// ---------------------------------------------------------------------------

interface ExportMeta {
  model?: string;
  platform?: string;
  source_url?: string;
  title?: string;
}

function extractMetadata(doc: PFMDocument): ExportMeta {
  const meta: ExportMeta = {};
  if (doc.meta.model) meta.model = doc.meta.model;
  if (doc.meta.agent) meta.platform = doc.meta.agent;
  if (doc.meta.source_url) meta.source_url = doc.meta.source_url;
  if (doc.meta.title) meta.title = doc.meta.title;
  // Also check for platform key directly
  if (doc.meta.platform) meta.platform = doc.meta.platform;
  return meta;
}

function getSection(doc: PFMDocument, name: string): string | undefined {
  const section = doc.sections.find((s) => s.name === name);
  return section?.content;
}

// ---------------------------------------------------------------------------
// Format exporters
// ---------------------------------------------------------------------------

function exportOpenAI(
  turns: Turn[],
  meta: ExportMeta,
): { line: string; turnCount: number } {
  const platform = meta.platform || 'unknown';
  const model = meta.model || 'unknown';
  const systemMsg = `Conversation from ${platform} using ${model}`;

  const messages: Array<{ role: string; content: string }> = [
    { role: 'system', content: systemMsg },
  ];
  for (const turn of turns) {
    messages.push({ role: turn.role, content: turn.content });
  }

  return { line: JSON.stringify({ messages }), turnCount: turns.length };
}

function exportAlpaca(
  turns: Turn[],
  meta: ExportMeta,
): { lines: string[]; turnCount: number } {
  const lines: string[] = [];
  let i = 0;
  while (i < turns.length) {
    if (
      turns[i].role === 'user' &&
      i + 1 < turns.length &&
      turns[i + 1].role === 'assistant'
    ) {
      const entry: Record<string, unknown> = {
        instruction: turns[i].content,
        input: '',
        output: turns[i + 1].content,
      };
      const metaObj: Record<string, string> = {};
      if (meta.model) metaObj.model = meta.model;
      if (meta.platform) metaObj.platform = meta.platform;
      if (Object.keys(metaObj).length > 0) entry.metadata = metaObj;
      lines.push(JSON.stringify(entry));
      i += 2;
    } else {
      i += 1;
    }
  }
  return { lines, turnCount: lines.length };
}

function exportShareGPT(
  turns: Turn[],
): { line: string; turnCount: number } {
  const conversations: Array<{ from: string; value: string }> = [];
  for (const turn of turns) {
    conversations.push({
      from: turn.role === 'user' ? 'human' : 'gpt',
      value: turn.content,
    });
  }
  return { line: JSON.stringify({ conversations }), turnCount: turns.length };
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export type ExportFormat = 'openai' | 'alpaca' | 'sharegpt';

export function exportDocument(
  doc: PFMDocument,
  format: ExportFormat = 'openai',
): { lines: string[]; turnCount: number } {
  // Prefer chain section; fall back to content
  const chain = getSection(doc, 'chain');
  let turns: Turn[];

  if (chain) {
    turns = parseTurns(chain);
  } else {
    const content = getSection(doc, 'content') || '';
    if (content) {
      turns = [{ role: 'assistant', content }];
    } else {
      return { lines: [], turnCount: 0 };
    }
  }

  if (turns.length === 0) return { lines: [], turnCount: 0 };

  const meta = extractMetadata(doc);

  switch (format) {
    case 'openai': {
      const r = exportOpenAI(turns, meta);
      return { lines: [r.line], turnCount: r.turnCount };
    }
    case 'alpaca': {
      const r = exportAlpaca(turns, meta);
      return { lines: r.lines, turnCount: r.turnCount };
    }
    case 'sharegpt': {
      const r = exportShareGPT(turns);
      return { lines: [r.line], turnCount: r.turnCount };
    }
    default:
      throw new Error(`Unknown export format: ${format}. Use: openai, alpaca, sharegpt`);
  }
}

export function exportDocuments(
  docs: PFMDocument[],
  format: ExportFormat = 'openai',
): { lines: string[]; totalTurns: number } {
  const allLines: string[] = [];
  let totalTurns = 0;
  for (const doc of docs) {
    const { lines, turnCount } = exportDocument(doc, format);
    allLines.push(...lines);
    totalTurns += turnCount;
  }
  return { lines: allLines, totalTurns };
}

export function loadPfmPaths(pathArg: string): string[] {
  const stat = statSync(pathArg);
  if (stat.isFile()) {
    return [pathArg];
  } else if (stat.isDirectory()) {
    return readdirSync(pathArg)
      .filter((f) => extname(f).toLowerCase() === '.pfm')
      .sort()
      .map((f) => join(pathArg, f));
  } else {
    throw new Error(`Path not found: ${pathArg}`);
  }
}

export function loadAndExport(
  pathArg: string,
  format: ExportFormat = 'openai',
): { lines: string[]; totalTurns: number; fileCount: number } {
  const paths = loadPfmPaths(pathArg);
  const docs: PFMDocument[] = paths.map((p) => {
    const text = readFileSync(p, 'utf-8');
    return parse(text);
  });
  const { lines, totalTurns } = exportDocuments(docs, format);
  return { lines, totalTurns, fileCount: paths.length };
}

File: pfm/export.py
"""
PFM Export - Turn parser and training data exporters.

Converts .pfm conversation files into fine-tuning JSONL formats:
  - OpenAI  (messages array with system/user/assistant roles)
  - Alpaca  (instruction/input/output per turn pair)
  - ShareGPT (conversations array with human/gpt roles)

Usage:
    from pfm.export import export_documents

    docs = [PFMReader.read(p) for p in pfm_paths]
    lines, total_turns = export_documents(docs, format="openai")
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Sequence

from pfm.document import PFMDocument
from pfm.reader import PFMReader


# ---------------------------------------------------------------------------
# Turn parser
# ---------------------------------------------------------------------------

def parse_turns(text: str) -> list[tuple[str, str]]:
    """Parse a chain string into (role, content) turn pairs.

    Chrome extension writes chains as:
        User: ...\n\nAssistant: ...

    We split on double-newline boundaries and detect User:/Assistant: prefixes.
    """
    turns: list[tuple[str, str]] = []
    if not text or not text.strip():
        return turns

    blocks = text.split("\n\n")
    current_role: str | None = None
    current_lines: list[str] = []

    for block in blocks:
        stripped = block.strip()
        if not stripped:
            continue

        # Detect role prefix
        if stripped.startswith("User:"):
            # Flush previous
            if current_role is not None:
                turns.append((current_role, "\n\n".join(current_lines).strip()))
            current_role = "user"
            current_lines = [stripped[len("User:"):].strip()]
        elif stripped.startswith("Assistant:"):
            if current_role is not None:
                turns.append((current_role, "\n\n".join(current_lines).strip()))
            current_role = "assistant"
            current_lines = [stripped[len("Assistant:"):].strip()]
        elif stripped.startswith("Agent:"):
            if current_role is not None:
                turns.append((current_role, "\n\n".join(current_lines).strip()))
            current_role = "assistant"
            current_lines = [stripped[len("Agent:"):].strip()]
        else:
            # Continuation of current block
            if current_role is not None:
                current_lines.append(stripped)
            else:
                # No role prefix yet — treat as assistant content
                current_role = "assistant"
                current_lines = [stripped]

    # Flush last turn
    if current_role is not None:
        turns.append((current_role, "\n\n".join(current_lines).strip()))

    return turns


def _extract_metadata(doc: PFMDocument) -> dict[str, str]:
    """Extract optional metadata fields from a PFM document."""
    meta: dict[str, str] = {}
    if doc.model:
        meta["model"] = doc.model
    if doc.agent:
        meta["platform"] = doc.agent
    # Check custom_meta for extra fields
    for key in ("source_url", "title", "platform"):
        val = doc.custom_meta.get(key, "")
        if val:
            meta[key] = val
    return meta


# ---------------------------------------------------------------------------
# Format exporters — each returns (json_line, turn_count)
# ---------------------------------------------------------------------------

def _export_openai(doc: PFMDocument, turns: list[tuple[str, str]], meta: dict[str, str]) -> tuple[str, int]:
    """Export one document as OpenAI fine-tuning format."""
    platform = meta.get("platform", "unknown")
    model = meta.get("model", "unknown")
    system_msg = f"Conversation from {platform} using {model}"

    messages: list[dict[str, str]] = [{"role": "system", "content": system_msg}]
    for role, content in turns:
        messages.append({"role": role, "content": content})

    return json.dumps({"messages": messages}, ensure_ascii=False), len(turns)


def _export_alpaca(doc: PFMDocument, turns: list[tuple[str, str]], meta: dict[str, str]) -> tuple[list[str], int]:
    """Export one document as Alpaca format (one line per user/assistant pair)."""
    lines: list[str] = []
    i = 0
    while i < len(turns):
        role, content = turns[i]
        if role == "user" and i + 1 < len(turns) and turns[i + 1][0] == "assistant":
            entry = {
                "instruction": content,
                "input": "",
                "output": turns[i + 1][1],
            }
            if meta:
                entry["metadata"] = meta
            lines.append(json.dumps(entry, ensure_ascii=False))
            i += 2
        else:
            i += 1
    return lines, len(lines)


def _export_sharegpt(doc: PFMDocument, turns: list[tuple[str, str]], meta: dict[str, str]) -> tuple[str, int]:
    """Export one document as ShareGPT format."""
    conversations: list[dict[str, str]] = []
    for role, content in turns:
        gpt_role = "human" if role == "user" else "gpt"
        conversations.append({"from": gpt_role, "value": content})

    return json.dumps({"conversations": conversations}, ensure_ascii=False), len(turns)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def export_document(doc: PFMDocument, fmt: str = "openai") -> tuple[list[str], int]:
    """Export a single PFM document to JSONL lines.

    Returns (lines, turn_count).
    """
    # Prefer chain section; fall back to content
    chain = doc.chain
    if chain:
        turns = parse_turns(chain)
    else:
        content = doc.content or ""
        if content:
            turns = [("assistant", content)]
        else:
            return [], 0

    if not turns:
        return [], 0

    meta = _extract_metadata(doc)

    if fmt == "openai":
        line, count = _export_openai(doc, turns, meta)
        return [line], count
    elif fmt == "alpaca":
        lines, count = _export_alpaca(doc, turns, meta)
        return lines, count
    elif fmt == "sharegpt":
        line, count = _export_sharegpt(doc, turns, meta)
        return [line], count
    else:
        raise ValueError(f"Unknown export format: {fmt!r}. Use: openai, alpaca, sharegpt")


def export_documents(docs: Sequence[PFMDocument], fmt: str = "openai") -> tuple[list[str], int]:
    """Export multiple PFM documents. Returns (all_lines, total_turns)."""
    all_lines: list[str] = []
    total_turns = 0
    for doc in docs:
        lines, count = export_document(doc, fmt)
        all_lines.extend(lines)
        total_turns += count
    return all_lines, total_turns


def load_pfm_paths(path: str) -> list[Path]:
    """Resolve a path to a list of .pfm file paths.

    If path is a file, returns [path].
    If path is a directory, returns all .pfm files in it (non-recursive).
    """
    p = Path(path)
    if p.is_file():
        return [p]
    elif p.is_dir():
        files = sorted(p.glob("*.pfm"))
        return files
    else:
        raise FileNotFoundError(f"Path not found: {path}")
